core/trino-main/src/main/java/io/trino/execution/scheduler/policy/SequentialExecutionSchedule.java

```diff
@@ -0,0 +1,606 @@
+        return state == SCHEDULED || state == RUNNING || state == FLUSHING || state.isDone();
+    }
+
+    private Set<PlanFragmentId> extractDependencies(Collection<PlanFragment> fragments)
+    {
+        // Build a graph where the plan fragments are vertexes and the edges represent
+        // a before -> after relationship. Destination fragment should be started only
+        // when source fragment is completed. For example, a join hash build has an edge
+        // to the join probe.
+        fragments.forEach(fragment -> {
+            fragmentDependency.addVertex(fragment.getId());
+            fragmentTopology.addVertex(fragment.getId());
+        });
+
+        Set<PlanFragmentId> nonLazyFragments = new HashSet<>();
+        Visitor visitor = new Visitor(fragments, fragmentDependency, fragmentTopology, nonLazyFragments);
+        fragments.forEach(fragment -> visitor.processFragment(fragment.getId()));
+
+        // Make sure there are no strongly connected components as it would mean circular dependency between stages
+        List<Set<PlanFragmentId>> components = new StrongConnectivityInspector<>(fragmentDependency).stronglyConnectedSets();
+        verify(components.stream().allMatch(set -> set.size() == 1), "circular dependency between stages");
+
+        return nonLazyFragments;
+    }
+
+    private static class Visitor
+            extends PlanVisitor<Fragments, PlanFragmentId>
+    {
+        private final Map<PlanFragmentId, PlanFragment> fragments;
+        private final DirectedGraph<PlanFragmentId, FragmentsEdge> fragmentDependency;
+        private final DirectedGraph<PlanFragmentId, FragmentsEdge> fragmentTopology;
+        private final Set<PlanFragmentId> nonLazyFragments;
+        private final Map<PlanFragmentId, Fragments> fragmentSources = new HashMap<>();
+
+        public Visitor(
+                Collection<PlanFragment> fragments,
+                DirectedGraph<PlanFragmentId, FragmentsEdge> fragmentDependency,
+                DirectedGraph<PlanFragmentId, FragmentsEdge> fragmentTopology,
+                Set<PlanFragmentId> nonLazyFragments)
+        {
+            this.fragments = requireNonNull(fragments, "fragments is null").stream()
+                    .collect(toImmutableMap(PlanFragment::getId, identity()));
+            this.fragmentDependency = requireNonNull(fragmentDependency, "fragmentDependency is null");
+            this.fragmentTopology = requireNonNull(fragmentTopology, "fragmentTopology is null");
+            this.nonLazyFragments = requireNonNull(nonLazyFragments, "nonLazyFragments is null");
+        }
+
+        public Fragments processFragment(PlanFragmentId planFragmentId)
+        {
+            if (fragmentSources.containsKey(planFragmentId)) {
+                return fragmentSources.get(planFragmentId);
+            }
+
+            Fragments fragment = processFragment(fragments.get(planFragmentId));
+            fragmentSources.put(planFragmentId, fragment);
```

Comment:
While visiting plan nodes `Fragments` does not contain currently visited stage. However, `processFragment` result does contain stage for which it was invoked. Maybe we could split `Fragments` into some intermediate structure (while visting plan nodes) and final one, but I don't think it would be simpler or easier to understand.